internal/datastore/proxy/relationshipintegrity.go

```diff
@@ -0,0 +1,397 @@
+	ds         datastore.Datastore
+	primaryKey hmacConfig
+	keysByID   map[string]hmacConfig
+}
+
+func (r *relationshipIntegrityProxy) lookupKey(keyID string) (hmacConfig, error) {
+	key, ok := r.keysByID[keyID]
+	if !ok {
+		return hmacConfig{}, fmt.Errorf("key not found: %s", keyID)
+	}
+
+	return key, nil
+}
+
+// computeRelationshipHash computes the HMAC hash of a relationship tuple.
+func computeRelationshipHash(tpl *corev1.RelationTuple, key hmacConfig) ([]byte, error) {
+	hmac := key.hmacPool.Get().(hash.Hash)
+	defer key.hmacPool.Put(hmac)
+	defer hmac.Reset()
+
+	var sb strings.Builder
+	sb.WriteString(tpl.ResourceAndRelation.Namespace)
+	sb.WriteString(":")
+	sb.WriteString(tpl.ResourceAndRelation.ObjectId)
+	sb.WriteString("#")
+	sb.WriteString(tpl.ResourceAndRelation.Relation)
+	sb.WriteString("@")
+	sb.WriteString(tpl.Subject.Namespace)
+	sb.WriteString(":")
+	sb.WriteString(tpl.Subject.ObjectId)
+	sb.WriteString("#")
+	sb.WriteString(tpl.Subject.Relation)
+
+	if tpl.Caveat != nil && tpl.Caveat.CaveatName != "" {
+		sb.WriteString(" with ")
+		sb.WriteString(tpl.Caveat.CaveatName)
+		sb.WriteString(":")
+		sb.WriteString(tpl.Caveat.Context.String())
+	}
+
+	if _, err := hmac.Write([]byte(sb.String())); err != nil {
+		return nil, err
+	}
+	return hmac.Sum(nil)[:16], nil
+}
+
+func (r *relationshipIntegrityProxy) SnapshotReader(rev datastore.Revision) datastore.Reader {
+	return relationshipIntegrityReader{
+		parent:  r,
+		wrapped: r.ds.SnapshotReader(rev),
+	}
+}
+
+func (r *relationshipIntegrityProxy) ReadWriteTx(ctx context.Context, f datastore.TxUserFunc, opts ...options.RWTOptionsOption) (datastore.Revision, error) {
+	return r.ds.ReadWriteTx(ctx, func(ctx context.Context, tx datastore.ReadWriteTransaction) error {
+		return f(ctx, &relationshipIntegrityTx{
+			ReadWriteTransaction: tx,
+			parent:               r,
+		})
+	}, opts...)
+}
+
+func (r *relationshipIntegrityProxy) CheckRevision(ctx context.Context, revision datastore.Revision) error {
+	return r.ds.CheckRevision(ctx, revision)
+}
+
+func (r *relationshipIntegrityProxy) Close() error {
+	return r.ds.Close()
+}
+
+func (r *relationshipIntegrityProxy) Features(ctx context.Context) (*datastore.Features, error) {
+	return r.ds.Features(ctx)
+}
+
+func (r *relationshipIntegrityProxy) HeadRevision(ctx context.Context) (datastore.Revision, error) {
+	return r.ds.HeadRevision(ctx)
+}
+
+func (r *relationshipIntegrityProxy) OptimizedRevision(ctx context.Context) (datastore.Revision, error) {
+	return r.ds.OptimizedRevision(ctx)
+}
+
+func (r *relationshipIntegrityProxy) ReadyState(ctx context.Context) (datastore.ReadyState, error) {
+	return r.ds.ReadyState(ctx)
+}
+
+func (r *relationshipIntegrityProxy) RevisionFromString(serialized string) (datastore.Revision, error) {
+	return r.ds.RevisionFromString(serialized)
+}
+
+func (r *relationshipIntegrityProxy) Statistics(ctx context.Context) (datastore.Stats, error) {
+	return r.ds.Statistics(ctx)
+}
```

Comment:
i like to make functions like these 1 single line so they all stack together without newlines between. it makes it more clear that they're just boilerplate

Reply:
Tried but go is reformatting it back to this

Reply:
Likely because the receiver is named `relationshipIntegrityProxy` and thus its making the lines too long. Shall I rename the struct?